File: 0454_med_hot_4sum-ii/solution.py
from collections import defaultdict
from typing import List

class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        records = defaultdict(int)

        n = len(nums1)

        for a in nums1:
            for b in nums2:
                records[a+b] += 1

        result = 0

        for c in nums3:
            for d in nums4:
                result += records[-(c + d)]
        
        return result

File: 0454_med_hot_4sum-ii/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        Map<Integer, Integer> records = new HashMap<>();

        for (int a : nums1) {
            for (int b : nums2) {
                records.put(a+b, records.getOrDefault(a+b, 0)+1);
            }
        }

        int result = 0;

        for (int c : nums3) {
            for ( int d : nums4) {
                result += records.getOrDefault(-(c+d), 0);
                
            }
        }

        return result;
        
    }
}